erigon-lib/commitment/hex_patricia_hashed.go

```diff
@@ -1998,6 +2084,104 @@ func (hph *HexPatriciaHashed) GenerateWitness(ctx context.Context, updates *Upda
 	return witnessTrie, rootHash, nil
 }
 
+func (phph *ParallelPatriciaHashed) Process(ctx context.Context, updates *Updates, logPrefix string) (rootHash []byte, err error) {
+	var (
+		ki uint64
+
+		updatesCount = updates.Size()
+		start        = time.Now()
+		logEvery     = time.NewTicker(20 * time.Second)
+	)
+	defer logEvery.Stop()
+	//hph.trace = true
+	nibbleUpdates := make([]*nibbleUpdate, 0)
+	currentNibble := 16 // invalid nibble value
+	errChan := make(chan error, 16)
+	nibblesBeingProcessed := 0
+
+	runParallelUnfolding := func() {
+		// we must submit the nibble updates and reinitialise slice
+		if len(nibbleUpdates) != 0 {
+			passedUpdates := append([]*nibbleUpdate{}, nibbleUpdates...)
+			/* go */ phph.StartKeyProcessing(ctx, errChan, passedUpdates)
```

Comment:
i would advise: to run group of "similar" goroutines: use `errgroup.WithContext`
it will:
- cancel other goroutines if one failed
- don't need `errChan` - just return `err` as any other golang code
- can set max amount of simultaneous goroutines (to limit ram use, or if machine is small, etc...): `g.SetLimit(6)` (even if you always calc 16 prefixes in-parallel)

Only important limitation: 
- use 1 `errgroup` for 1 set of similar goroutines. Example: If you have another goroutine for logging progress - don't add it to `errgroup`. 

-